src/controllers/dynakube/activegate/reconciler.go

```diff
@@ -2,263 +2,101 @@ package activegate
-	if updated || err != nil {
-		return updated, errors.WithStack(err)
-	}
-
-	return false, nil
-}
-
-func (r *Reconciler) buildDesiredStatefulSet() (*appsv1.StatefulSet, error) {
-	kubeUID, err := kubesystem.GetUID(r.apiReader)
-	if err != nil {
-		return nil, errors.WithStack(err)
-	}
-
-	activeGateConfigurationHash, err := r.calculateActiveGateConfigurationHash()
-	if err != nil {
-		return nil, errors.WithStack(err)
-	}
-
-	stsProperties := statefulset.NewStatefulSetProperties(
-		r.Instance, r.capability, kubeUID, activeGateConfigurationHash, r.feature, r.capabilityName, r.serviceAccountOwner,
-		r.initContainersTemplates, r.containerVolumeMounts, r.volumes)
-	stsProperties.OnAfterCreateListener = r.onAfterStatefulSetCreateListener
-
-	desiredSts, err := statefulset.CreateStatefulSet(stsProperties)
-	return desiredSts, errors.WithStack(err)
-}
-
-func (r *Reconciler) getStatefulSet(desiredSts *appsv1.StatefulSet) (*appsv1.StatefulSet, error) {
-	var sts appsv1.StatefulSet
-	err := r.Get(context.TODO(), client.ObjectKey{Name: desiredSts.Name, Namespace: desiredSts.Namespace}, &sts)
-	if err != nil {
-		return nil, errors.WithStack(err)
-	}
-	return &sts, nil
-}
-
-func (r *Reconciler) createStatefulSetIfNotExists(desiredSts *appsv1.StatefulSet) (bool, error) {
-	_, err := r.getStatefulSet(desiredSts)
-	if err != nil && k8serrors.IsNotFound(errors.Cause(err)) {
-		log.Info("creating new stateful set for " + r.feature)
-		return true, r.Create(context.TODO(), desiredSts)
-	}
-	return false, err
-}
-
-func (r *Reconciler) updateStatefulSetIfOutdated(desiredSts *appsv1.StatefulSet) (bool, error) {
-	currentSts, err := r.getStatefulSet(desiredSts)
-	if err != nil {
+	if err := r.reconcileActiveGateProxySecret(); err != nil {
 		return false, err
 	}
-	if !kubeobjects.HasChanged(currentSts, desiredSts) {
-		return false, nil
-	}
-
-	if kubeobjects.LabelsNotEqual(currentSts.Spec.Selector.MatchLabels, desiredSts.Spec.Selector.MatchLabels) {
-		return r.recreateStatefulSet(currentSts, desiredSts)
-	}
 
-	log.Info("updating existing stateful set")
-	if err = r.Update(context.TODO(), desiredSts); err != nil {
-		return false, err
+	var caps = capability.GenerateActiveGateCapabilities(r.Dynakube)
+	for _, agCapability := range caps {
+		if agCapability.Enabled() {
+			return r.createCapability(agCapability)
+		} else {
+			if err := r.deleteCapability(agCapability); err != nil {
+				return false, err
+			}
+		}
 	}
 	return true, err
 }
 
-func (r *Reconciler) recreateStatefulSet(currentSts, desiredSts *appsv1.StatefulSet) (bool, error) {
-	log.Info("immutable section changed on statefulset, deleting and recreating", "name", desiredSts.Name)
-
-	err := r.Delete(context.TODO(), currentSts)
-	if err != nil {
-		return false, err
+func (r *Reconciler) reconcileActiveGateProxySecret() (err error) {
+	gen := capability.NewActiveGateProxySecretGenerator(r.Client, r.apiReader, r.Dynakube.Namespace, log)
```

Comment:
introduces a dependency on a secret generator instance which cannot be mocked / faked when testing